salt/crypt.py

```diff
@@ -126,13 +138,23 @@ def gen_keys(keydir, keyname, keysize, user=None, passphrase=None):
     if not os.access(keydir, os.W_OK):
         raise IOError('Write access denied to "{0}" for user "{1}".'.format(os.path.abspath(keydir), getpass.getuser()))
 
-    cumask = os.umask(191)
-    with salt.utils.files.fopen(priv, 'wb+') as f:
-        f.write(gen.exportKey('PEM', passphrase))
+    cumask = os.umask(0o277)
+    if HAS_M2:
+        # if passphrase is empty or None use no cipher
+        if not passphrase:
+            gen.save_pem(priv, cipher=None)
+        else:
+            gen.save_pem(priv, cipher='des_ede3_cbc', callback=lambda x: six.b(passphrase))
```

Comment:
Is this the cipher used by the pycrypto lib? Can't the encryption type be autodetected so that we could use AES here?

Reply:
Yes this is used by pycrypto by default. We could use autodetect here, I think. But I need to have a review to say exactly

Reply:
I am ok with this otherwise because this is not a default setting. If anything to just get this into the codebase.
But we should look at autodetect here in a future PR

Reply:
Salt requires pycrypto>=2.6.1, but pycrypto 2.6.1 doesn't support AES algorithm. So it looks to better to make this change later when user will have a way to choose what cryptographic backend to use to not be surprised.

Reply:
But yes, it's a very good catch, it should be changed to autodetect, or configurable in the future.

Reply:
Thanks fro the research, again, I am ok with where we are here do far and I think that we can move forward, the password protected key here is not default and we can make this configurable in the future